lib/util/utils.go

```diff
@@ -170,3 +174,69 @@ func Address(network, host string) string {
 	}
 	return u.String()
 }
+
+// AsService wraps the given function to implement suture.Service by calling
+// that function on serve and closing the passed channel when Stop is called.
+func AsService(fn func(stop chan struct{})) suture.Service {
+	return AsServiceWithError(func(stop chan struct{}) error {
+		fn(stop)
+		return nil
+	})
+}
+
+type ServiceWithError interface {
+	suture.Service
+	Error() error
+}
+
+// AsServiceWithError does the same as AsService, except that it keeps track
+// of an error returned by the given function.
+func AsServiceWithError(fn func(stop chan struct{}) error) ServiceWithError {
+	s := &service{
+		serve:   fn,
+		stop:    make(chan struct{}),
+		stopped: make(chan struct{}),
+		mut:     sync.NewMutex(),
+	}
+	close(s.stopped) // not yet started, don't block on Stop()
+	return s
+}
+
+type service struct {
+	serve    func(stop chan struct{}) error
+	stopping bool
+	stop     chan struct{}
+	stopped  chan struct{}
+	err      error
+	mut      sync.Mutex
+}
+
+func (s *service) Serve() {
+	s.mut.Lock()
+	if s.stopping {
+		s.mut.Unlock()
+		return
+	}
+	s.err = nil
+	s.stopped = make(chan struct{})
+	s.mut.Unlock()
+	defer close(s.stopped)
+	err := s.serve(s.stop)
+	s.mut.Lock()
+	s.err = err
+	s.mut.Unlock()
+}
+
+func (s *service) Stop() {
+	s.mut.Lock()
+	s.stopping = true
+	s.mut.Unlock()
+	close(s.stop)
+	<-s.stopped
```

Comment:
I think these services can never be restarted, as s.stop is only initialised once?
Also it feels wrong touching s.stop and s.stopped outside of the lock, as they are clearly initialised under the lock, so I feel this is racey.

Reply:
They indeed can't be restarted after `Stop`, that's by design as mandated by suture:

> [...] A supervisor will call .Stop() only once. Thus, it may be as destructive as it likes to get the service to stop.
> 
> Once Stop has been called on a Service, the Service SHOULD NOT be reused in any other supervisor! [...]

I added `stopping` after a race came up in testing - in hindsight that's an ugly bolt-on. I revised the stop/stopped-locking-machinism, I hope it's clearer now (and should be race free).